From d22b41d8c1db190502a0da463a145e72b5be7d4e Mon Sep 17 00:00:00 2001
From: LaserEyess <lasereyess@users.noreply.github.com>
Date: Fri, 21 Oct 2022 20:08:23 -0400
Subject: [PATCH] context_drm_egl: add colorspace_hints swapchain function

This adds the ability to use colorspace hints in the drm context for
opengl. This will be useful for sending HDR metadata to the screen.
---
 video/out/opengl/context.c         |  2 ++
 video/out/opengl/context_drm_egl.c | 22 +++++++++++++++++++---
 2 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/video/out/opengl/context.c b/video/out/opengl/context.c
index 377b678cde55..96925a47a6db 100644
--- a/video/out/opengl/context.c
+++ b/video/out/opengl/context.c
@@ -150,6 +150,8 @@ bool ra_gl_ctx_init(struct ra_ctx *ctx, GL *gl, struct ra_gl_ctx_params params)
     if (ext) {
         if (ext->color_depth)
             p->fns.color_depth = ext->color_depth;
+        if (ext->colorspace_hint)
+            p->fns.colorspace_hint = ext->colorspace_hint;
         if (ext->start_frame)
             p->fns.start_frame = ext->start_frame;
         if (ext->submit_frame)
diff --git a/video/out/opengl/context_drm_egl.c b/video/out/opengl/context_drm_egl.c
index 022a2d2a4551..350c3d82e8f2 100644
--- a/video/out/opengl/context_drm_egl.c
+++ b/video/out/opengl/context_drm_egl.c
@@ -103,6 +103,8 @@ struct priv {
 
     struct mpv_opengl_drm_params_v2 drm_params;
     struct mpv_opengl_drm_draw_surface_size draw_surface_size;
+
+    struct mp_colorspace out_color;
 };
 
 // Not general. Limited to only the formats being used in this module
@@ -645,10 +647,24 @@ static void drm_egl_swap_buffers(struct ra_swapchain *sw)
     }
 }
 
+static bool drm_egl_colorspace_hint(struct ra_swapchain *sw,
+                                    const struct mp_image *csp_hint_frame,
+                                    struct mp_colorspace *configured_csp)
+{
+    struct priv *p = sw->ctx->priv;
+
+    if (csp_hint_frame) {
+        p->out_color = csp_hint_frame->params.color;
+        return true;
+    }
+    return false;
+}
+
 static const struct ra_swapchain_fns drm_egl_swapchain = {
-    .start_frame   = drm_egl_start_frame,
-    .submit_frame  = drm_egl_submit_frame,
-    .swap_buffers  = drm_egl_swap_buffers,
+    .colorspace_hint = drm_egl_colorspace_hint,
+    .start_frame     = drm_egl_start_frame,
+    .submit_frame    = drm_egl_submit_frame,
+    .swap_buffers    = drm_egl_swap_buffers,
 };
 
 static void drm_egl_uninit(struct ra_ctx *ctx)
